pkgs/by-name/tp/tpm2-pkcs11/package.nix

```diff
@@ -1,109 +1,195 @@
 
-  configureFlags = lib.optionals (!fapiSupport) [
-    # Note: this will be renamed to with-fapi in next release.
-    "--enable-fapi=no"
-  ];
+  configureFlags =
+    [
+      (lib.enableFeature finalAttrs.doCheck "unit")
+      (lib.enableFeature finalAttrs.doCheck "integration")
+    ]
+    ++ lib.optionals enableFuzzing [
+      "--enable-fuzzing"
+      "--disable-hardening"
+    ]
+    ++ lib.optional fapiSupport "--with-fapi";
+
+  strictDeps = true;
 
   nativeBuildInputs = [
-    pkg-config
-    autoreconfHook
     autoconf-archive
+    autoreconfHook
     makeWrapper
     patchelf
-  ];
-  buildInputs = [
-    tpm2-tss
-    tpm2-tools
-    opensc
-    openssl
-    sqlite
-    libyaml
+    pkg-config
     (python3.withPackages (
-      ps: with ps; [
+      ps:
+      with ps;
+      [
         packaging
         pyyaml
+        python-pkcs11
         cryptography
         pyasn1-modules
         tpm2-pytss
       ]
+      ++ cryptography.optional-dependencies.ssh
     ))
   ];
 
+  buildInputs = [
+    libyaml
+    opensc
+    openssl
+    sqlite
+    tpm2-tools
+    tpm2-tss
+  ];
+
+  nativeCheckInputs = [
+    dbus
+    expect
+    gnutls
+    iproute2
+    nss.tools
+    opensc
+    openssh
+    openssl
+    p11-kit
+    sqlite
+    swtpm
+    tpm2-abrmd
+    tpm2-tools
+    which
+    xxd
+  ];
+
+  checkInputs = [
+    cmocka
+    tpm2-abrmd
+  ];
+
+  enableParallelBuilding = true;
+  hardeningDisable = lib.optional enableFuzzing "all";
+
   outputs = [
     "out"
     "bin"
     "dev"
   ];
 
+  doCheck = true;
   dontStrip = true;
   dontPatchELF = true;
 
-  # To be able to use the userspace resource manager, the RUNPATH must
-  # explicitly include the tpm2-abrmd shared libraries.
-  preFixup =
+  preConfigure =
+    let
+      ldflags = [
+        "-Wl,--push-state,--no-as-needed"
+        "-ltss2-tcti-device"
+        "-ltss2-tcti-tabrmd"
+        "-Wl,--pop-state"
+      ];
+    in
+    lib.optionalString abrmdSupport ''
+      configureFlagsArray+=(EXTRA_LDFLAGS=${lib.escapeShellArg ldflags})
+    '';
+
+  preCheck =
     let
-      rpath = lib.makeLibraryPath (
-        (lib.optional abrmdSupport tpm2-abrmd)
-        ++ [
-          tpm2-tss
-          sqlite
+      openssl-modules = buildEnv {
+        name = "openssl-modules";
+        pathsToLink = [ "/lib/ossl-modules" ];
+        paths = map lib.getLib [
           openssl
-          glibc
-          libyaml
-        ]
-      );
+          tpm2-openssl
+        ];
+      };
     in
     ''
-      patchelf \
-        --set-rpath ${rpath} \
-        ${lib.optionalString abrmdSupport "--add-needed ${lib.makeLibraryPath [ tpm2-abrmd ]}/libtss2-tcti-tabrmd.so"} \
-        --add-needed ${lib.makeLibraryPath [ tpm2-tss ]}/libtss2-tcti-device.so \
-        $out/lib/libtpm2_pkcs11.so.0.0.0
+      # Enable tests to load TCTI modules
+      export LD_LIBRARY_PATH+=":${
+        lib.makeLibraryPath [
+          swtpm
+          tpm2-abrmd
+          tpm2-tools
+        ]
```

Comment:
```suggestion
        lib.makeLibraryPath [
          swtpm
          tpm2-tools
        ] ++ lib.optionals (!abrmdSupport) [
          # Only add this if userspace‐resource manager support has been disabled to check
          # during testing whether it has been properly linked at compile time during.
          tpm2-abrmd
        ]
```

Reply:
The passthru tests actually fail if I do that

Reply:
Maybe my approach to inject the linker flags was incorrect. If in doubt, just revert to the previous variant using `patchelf`.